Use Python stack for scraper

EthShaw opened the pull request:
Rewrote the gawk/bash scraper scripts in Python. I updated the Readme with the new changes. When rewriting the README, I noticed that the TBA-auth file is not supposed to be included in the repository. Since it makes sense that the TBA key should not be redistributed, I removed it from the repository and added some instructions explaining an overview of how to get one from thebluealliance.com. I also made sure the program creates a descriptive warning if this file does not exist. The new scraper.py replaces all of the old scripts except ask_google, which I did not change.

Fixes #7

ErikBoesen commented on README.md:
```diff
@@ -0,0 +1,66 @@
+
+## Descriptions of Files
+
+**README.md** - This file.
```
I feel like you can probably just omit this

ErikBoesen commented on README.md:
```diff
@@ -0,0 +1,66 @@
+**[YEAR](YEAR)** - A file containing the current year for purposes of
+retrieving team lists from thebluealliance.com.
+
+**[explanation of process.md](explanation%20of%20process.md)** - This file
```
Can we merge this into this README? Or at the very least, give it a name without spaces?

EthShaw replied:
Honestly, I don't know why I included the file. I mostly made it when dissecting the original gawk script so it would be easier to rewrite it. I could just delete it and merge the part that pertains to the current version into the README.

ErikBoesen replied:
That would be ideal!

ErikBoesen commented:
Amazing work. Thank you so much.

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+    global useCache
+    print ('Downloading GeoNames data...')
+
+    for file in POSTAL_FILES:
```
Consider expanding this, such as:
```py
for url, filename, needs_unzip in POSTAL_FILES:
    ...
```
(You'll need to adjust if you use my array suggestion.)

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+
+# What files to download, name to save them as (within CACHE_DIR), and whether or not they need to be unzipped.
+POSTAL_FILES = [
+    [ 'https://download.geonames.org/export/zip/allCountries.zip',       'allCountries.zip',     True  ],
```
Alternatively, at least make them tuples.

ErikBoesen replied:
tbh, if you wanted to make a class for these that could also be neat.

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+                req.raise_for_status()
+
+                with open(path, 'wb') as writer:
+                    for chunk in req.iter_content(chunk_size=16384):
```
Can we make a constant for this chunk size rather than the magic number?

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+def load_geonames_data():
+    print('Loading GeoNames data...')
+
+    """ Definitions """
```
Weird comment format

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+    # reads the file line by line (row by row) and breaks each row into
+    # columns, calling callback with each row (as a list of columns). The sep
+    # parameter changes the separator, so it can read CSV files, etc. as well.
+    def readTSV(file, callback, sep='\t'):
```
Snake case again

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+                # Skip empty and blank lines and comments (which start with #)
+                if line.lstrip().startswith('#') or not line or line.isspace():
+                    continue
+                row = line.split(sep)
```
Do you want to strip this before splitting it?

EthShaw replied:
The original scraper doesn't strip the lines before splitting, so as far as I am aware it isn't necessary. The reason for the lstrip before the # check is in case any of the lines with comments have leading spaces. In the original scraper, comments were just treated like all the other lines, but lines that didn't have enough columns after splitting on \t were ignored, which effectively eliminated the comments since none of them contained tabs. Do you think I should strip the lines before splitting them?

ErikBoesen replied:
I misremembered how `.split()` works, disregard.

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+                callback(row)
+
+
+    """ Initialize the country codes table """
```
Again, would prefer # style comment

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+
+
+    """ Initialize the country codes table """
+    print ('Loading country code mappings...')
```
Remove the space between `print` and `(`

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+    geoNames['ccodes'].pop('', None)
+
+
+    """
```
Comment styling

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+
+
+    """
+        Initialize the cities dictionary, which holds the lattitude and
```
*latitude

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+    # Load the lattitude and longitude for each city in cities1000.txt and put
+    # them in geoNames['cities']
+    def proccessCities(row):
+        cityNameAscii = row[2].upper()
```
Snake case for all this stuff

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+        if not adminNameAscii:
+            return
+        
+        # This helper function puts the lattitude and longitude of the city
```
*latitude

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+            province = 'SANTIAGO METROPOLITAN'
+
+        if countryCode == 'CN' and province == 'HUNAN':
+            province = 'HENAN'
```
❗  Hunan and Henan are two separate provinces.

EthShaw replied:
Interesting. I wonder why this was in the original script. It doesn't seem to break anything when I remove it.

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+            elif province == 'PA' and city == 'WARMINSTER':
+                city = 'WARMINSTER HEIGHTS'
+            elif province =='MO' and city == 'LEES SUMMIT':
+                city = "LEE'S SUMMIT"
```
I generally prefer sticking with the standardized single quotes and escaping the apostrophe, but that's just me.

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+                city = 'NEW YORK CITY'
+            elif province == 'PA' and city == 'WARMINSTER':
+                city = 'WARMINSTER HEIGHTS'
+            elif province =='MO' and city == 'LEES SUMMIT':
```
Missing a space between `==` and `'MO'`

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+            province = 'NORTH BRABANT'
+
+        if countryCode == 'DO' and province == 'SANTO DOMINGO' and city == province:
+            province = 'NACIONAL'
```
Could there be a better way to specify these replacements? Like a JSON file that can specify in a standard format so you can iterate through and do all the replacements without having to hardcode them? Just a thought, not a blocker.

EthShaw replied:
I might address this in a separate PR.

ErikBoesen replied:
That's fine.

ErikBoesen reviewed (changes requested):
I know this is a lot of comments, but this is still excellent work, and most of these complaints are quite minor.

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+print(f'Downloaded data for year {YEAR}')
+
+# The directory where downloaded GeoNames data is cached
+CACHE_DIR = Path.cwd() / 'cache'
```
This isn't an issue, more just a comment as I'm unfamiliar with this pretty neat syntax

EthShaw replied:
Yes, this is valid syntax. The `pathlib` library overrides the `/` operator per [https://docs.python.org/3/library/pathlib.html#operators](https://docs.python.org/3/library/pathlib.html#operators). I don't have a lot of experience with Python, so I found this when I was looking at how to do file and path manipulation. It is a pretty neat and concise syntax.

ErikBoesen replied:
Very neat!

EthShaw commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+# Additional country code mappings. Currently, this adds country
+# codes for Czech Republic (CZ) and Chinese Taipei (TW) because
+# those are the names used by FIRST/TBA, but Geonames has those
+# countries listed as Czechia Prague and Taiwan, respectively.
```
As it turns out, my comment is actually incorrect. I thought it was strange, and I looked back and realized I must have mistaken a tab for a space (it was only 1 character width because of the alignment). It's actually only listed as "Czechia" in the GeoNames database.

EthShaw commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+
+# Load the auth key
+AUTH_PATH = Path('TBA-auth')
+if not AUTH_PATH.exists():
```
What are the differences between `os.path.exists` and `pathlib`'s `Path.exists`?

ErikBoesen replied:
Probably not much. Since you've standardized on using `pathlib` in this script, probably fine to stick with that.

EthShaw commented:
I believe I addressed all of your comments. Just a note, the scraper indents `teams.json` with tabs. I chose this because it uses less bytes each line, but the original uses three spaces, so I can switch to that (or another number of spaces) if you would prefer.

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+
+    # Load the latitude and longitude for each city in cities1000.txt and put
+    # them in geoNames['cities']
+    def proccess_cities(row):
```
*process

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+        state_code = row[10]
+        # Name of the administrative division (state, province, etc.)
+        # See comment above about the adms dictionary for further explanation.
+        admin_name_ascii = geo_names['adms'].get(
```
I don't think this needs to be on two lines

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+
+
+class GeoNamesFile():
+    """A helper class to hold the url and name of a GeoNames file as
```
For a multi-line docstring, the convention is to put the """s on separate lines from the content of the docstring.
```py
"""
A helper class
etc.
"""
```

ErikBoesen commented on scraper.py:
```diff
@@ -0,0 +1,613 @@
+    print('Loading administrative division names...')
+    geo_names['adms'] = {}
+
+    def proccess_admin_codes(row):
```
*process again here

ErikBoesen reviewed (changes requested):
Almost there.

ErikBoesen commented:
Thank you!
